Pixelation-Labs/pixelmon-contracts
# Documentation and unit test for WinnerSelection in pixelmon trainer adventure

This PR adds documentation and unit test for Pixelmon Trainer Adventure smart contract.

### How to test

To run the unit test, we need to change the smart contract to do the mock testing

Inside the **WinnerSelectionManager.sol**, replace the `generateChainLinkRandomNumbers` with the following code:

```sol
    uint256 reqId;
    /// @notice Generate random number from Chainlink
    /// @param _weekNumber Number of the week
    /// @return requestId Chainlink requestId
    function generateChainLinkRandomNumbers(
        uint256 _weekNumber
    ) external onlyModerator(msg.sender) validWinnerUpdationPeriod(_weekNumber) validWeekNumber(_weekNumber) returns (uint256 requestId) {

        requests[reqId] = Request({randomWords : new uint256[](0), exists : true, fulfilled : false, weekNumber : _weekNumber});
        requestId = COORDINATOR.requestRandomWords(keyHash, chainLinkSubscriptionId, requestConfirmations, callbackGasLimit, Random_Number_Count);
        //requests[requestId] = Request({randomWords: new uint256[](0), exists: true, fulfilled: false, weekNumber: _weekNumber});
        requestIds.push(requestId);
        lastRequestId = requestId;
        return requestId;
```

Inside the **Utils.sol** change the return value for `getRandomNumber` method to this

```
return 2;
```

After that we can start the test with

```
yarn test:coverage:revised:pixelmon:trainer:adventure
```

### Coverage

```
 RevisedPixelmonTrainerAdventure/                    |    99.38 |    93.42 |     97.3 |    99.07 |                |
  IPxTrainerAdventureSignature.sol                   |      100 |      100 |      100 |      100 |                |
  PxTrainerAdventure.sol                             |      100 |    95.83 |      100 |      100 |                |
  PxTrainerAdventureSignature.sol                    |      100 |      100 |      100 |      100 |                |
  Utils.sol                                          |    85.71 |       50 |    66.67 |       80 |          17,18 |
  WinnerSelectionManager.sol                         |      100 |    95.45 |      100 |      100 |                |
  ```
